### PR TITLE
Deprecate non-Pacemaker-specific public C APIs for 2.1.0

### PR DESCRIPTION
**Pull request by kgaillot**
(EDITED) This deprecates (via doxygen comments and PCMK__NO_COMPAT) a couple of macros in Pacemaker's public C API in libcrmcommon.

The overarching intent is that Pacemaker's public API should be Pacemaker-specific, and not provide generic capabilities. We don't want to have to support our own equivalent of glib, and we do want the flexibility to reimplement such interfaces as needed.

The plan is to deprecate most generic APIs in 2.1.0, and remove them from the public API in a future series (2.2.0 or 3.0.0 or whatever). The capabilities will remain as private API, as new functions that follow current guidelines for naming, return values, etc., with the old names as wrappers.

### CODE REVIEW

**Comment by kgaillot** [comment]
@clumens , can you review this when you get a chance? It looks big but most commits are one-liners

**Comment by kgaillot** [comment]
I went with separate commits for most items so that they show up as separate lines in the new "API changes" section of the ChangeLog coming in 2.1.0

**Comment by wenningerk** [comment]
This hits SBD in a couple of areas like logging, mainloop and single occasions GListPtr and crm_strdup_printf.
This is just the result of a quick manual scan.
Do we already have a simple automatic way to find deprecate API usage?
If not could we use the doxygen-comments to build a set of headers that doesn't expose the deprecated stuff?
Creating a version of the libraries that has the symbols removed wouldn't catch stuff going in via macros and simple defines.
Was thinking of a macro that can be used to mangle deprecated symbols (didn't we have such a thing already somewhere in pacemaker? Didn't find anything with a quick grep at least though ...) though that wouldn't be enough as stated.

**Comment by clumens** [comment]
> Do we already have a simple automatic way to find deprecate API usage?
> If not could we use the doxygen-comments to build a set of headers that doesn't expose the deprecated stuff?
> Creating a version of the libraries that has the symbols removed wouldn't catch stuff going in via macros and simple defines.

We could add `__attribute__((__deprecated__))` to the function prototypes, but that won't help with deprecated macros.  There doesn't really appear to be any good way of marking those as deprecated, aside from hacks like wrapping them in something like `#ifdef DEPRECATED_MACROS`.

File: include/crm/crm.h [hunk 1]
```diff
@@ -53,7 +53,14 @@ extern "C" {
  */
 #  define CRM_FEATURE_SET		"3.7.2"
 
+//!@{
+//! \deprecated This will be removed in a future release
+
 #  define EOS		'\0'
```
Comment (clumens) [inline include/crm/crm.h:59]: This is also defined in include/portability.h, and is only used at one spot in lib/common/strings.c.  So we could probably get rid of our use entirely and just leave the deprecated macros around for whoever might be using it.

**Comment by kgaillot** [comment]
> This hits SBD in a couple of areas like logging, mainloop and single occasions GListPtr and crm_strdup_printf.
> This is just the result of a quick manual scan.

Bummer, I assumed those were safe to drop, I should have grepped sbd.

I suppose it does make sense to keep crm_set_log_level() public API, so applications can control how chatty the libraries are. I tend to think crm_debug(), crm_perror() and friends should not be public API, and instead qb_log() etc. should be used directly if desired. Do the pacemaker wrappers provide anything besides convenience?

crm_strdup_printf() is exactly the sort of thing I don't want to have to maintain as public API; pacemaker shouldn't be a general-purpose utility library. GListPtr is even worse since pacemaker shouldn't be stepping on glib's namespace.

The mainloop stuff I'm ambivalent about. It's not pacemaker-specific in any way, so I'd rather not be tied down to a public interface, but the signal and trigger events are nice features GMainLoop doesn't have. I suppose we can keep it public.

I'll go through and remove anything from this PR that's being used by sbd currently. We can update sbd so it doesn't use the undesirable pieces and deprecate those in the next round in a few years, so we don't deprecate anything that is still used in recent distro versions. And of course it wouldn't be actually removed until much later than that.

> Do we already have a simple automatic way to find deprecate API usage?
> If not could we use the doxygen-comments to build a set of headers that doesn't expose the deprecated stuff?

We have a symbol PCMK__NO_COMPAT that pacemaker code defines, that ifdef's out certain deprecated stuff in headers. That allows us to make sure we don't use deprecated API in pacemaker while still leaving it available for third-party code. I wouldn't use it in sbd releases but it wouldn't hurt occasionally building with it defined just to see.

However we can't use that for all deprecated items. E.g. a lot of this PR is stuff that I want to deprecate externally but keep available internally. Such things are deprecated only via doxygen comments.

> Creating a version of the libraries that has the symbols removed wouldn't catch stuff going in via macros and simple defines.
> Was thinking of a macro that can be used to mangle deprecated symbols (didn't we have such a thing already somewhere in pacemaker? Didn't find anything with a quick grep at least though ...) though that wouldn't be enough as stated.

I suppose we could define something for the reverse of PCMK__NO_COMPAT, things that are deprecated externally but not internally. I doubt it's worth the effort though. My long-term goal is to define new internal API that conforms to current guidelines, and leave the old names as wrappers guarded by PCMK__NO_COMPAT.

**Comment by kgaillot** [comment]
> > Do we already have a simple automatic way to find deprecate API usage?
> > If not could we use the doxygen-comments to build a set of headers that doesn't expose the deprecated stuff?
> > Creating a version of the libraries that has the symbols removed wouldn't catch stuff going in via macros and simple defines.
> 
> We could add `__attribute__((__deprecated__))` to the function prototypes

That's an interesting idea, but it would break anything compiled with fatal warnings, so I'd lean against it. Unless we want to do something like deprecate via doxygen comments for a few years, then add the attribute for a few more years, then finally remove it.

> but that won't help with deprecated macros. There doesn't really appear to be any good way of marking those as deprecated, aside from hacks like wrapping them in something like `#ifdef DEPRECATED_MACROS`.

**Comment by clumens** [comment]
> That's an interesting idea, but it would break anything compiled with fatal warnings, so I'd lean against it. Unless we want to do something like deprecate via doxygen comments for a few years, then add the attribute for a few more years, then finally remove it.

This seems like a bit of an arms race, but we could add something like `-Wno-error=deprecated` to configure.ac where we test for fatal warnings.  I haven't tested this, but it looks like it should be doable.  This would still hit people that go out of their way to add -Werror on their own, of course.

**Comment by wenningerk** [comment]
I guess everything but the mainloop stuff can be removed right away from sbd.
And that is just part of the pacemaker-watcher where it actually makes sense
that it is looking a bit like pacemaker-code.
Haven't tried out if it is working out of the box - guess not as the closing clause is missing at file-ends iirc - but 
couldn't we simply replace the doxygen-deprecation start-clause by '#if 0' and the end-clause by '#endif'?
Probably doxygen-comments need to be checked/modified a bit but we wouldn't have to add anything that might
possibly break builds against pacemaker or clogs the code with ugly stuff.
Could package those headers in something like pacemaker-libs-deprecation-check-devel.
And it would work both for macros and symbols.

**Comment by kgaillot** [comment]
> Haven't tried out if it is working out of the box - guess not as the closing clause is missing at file-ends iirc - but
> couldn't we simply replace the doxygen-deprecation start-clause by '#if 0' and the end-clause by '#endif'?

That's exactly what PCMK__NO_COMPAT is for. This particular PR is unusual in that we're keeping the APIs available internally, so it doesn't work here. Probably the cleanest solution is to not deprecate anything until a new private API has replaced them internally, so we can use PCMK__NO_COMPAT. Now that we have coccinelle, that should easy in most cases, so maybe I'll go that route.

**Comment by wenningerk** [comment]
> That's exactly what PCMK__NO_COMPAT is for. This particular PR is unusual in that we're keeping the APIs available internally, so it doesn't work here. Probably the cleanest solution is to not deprecate anything until a new private API has replaced them internally, so we can use PCMK__NO_COMPAT. Now that we have coccinelle, that should easy in most cases, so maybe I'll go that route.

I know but when introducing PCMK__NO_COMPAT in addition wherever the doxygen-comments are that clutters the code and is just useful for such a deprecation test-build. Code would be turned on in the pacemaker-build and off for e.g. test-building sbd to see if deprecated API is still used. With creating an additional devel-rpm for doing such test builds that has the doxygen-comments replaced the code wouldn't be additionally cluttered. Not even build defines would have to be altered. Just switch the devel-rpm (e.g. in BuildRequires). Coccinelle might be useful for mangling the headers if more than a simple find/replace should be needed. (If one wanted to build both conflicting devel-rpms in one build that  might work the straight forward way out of obvious reasons. Would be interesting if it is possible to have a different buildroot for one of the sub-packages or some other way to handle this class of issues.)
When building with warnings as errors it should be enough to have the headers replaced as we anyway shouldn't get till linking if deprecated API is being used.
If it should be really needed that the linker does do the check on top one possibility would be to have all deprecated symbols altered by a macro within the pacemaker-build and when the headers are pulled in by the to be tested package this macro is turned off. Or even without rebuilding the dynamic linker wouldn't find the un-mangled symbols.

**Comment by kgaillot** [comment]
@clumens @wenningerk I've updated the PR with just a couple of deprecations to show a new approach.

The key change is moving the deprecated sections of existing public headers to new headers with "_compat.h" added. The original public header includes the compatibility header unless PCMK__NO_COMPAT is defined. This has a number of benefits -- clearly isolating deprecated code, allowing the source file that defines the old functions to include the compatibility header rather than redeclare everything, and putting the deprecated code on separate pages in the doxygen API documentation.

The other difference is that the old names will not continue to be used by internal code. We can define new names for anything we want to keep for internal purposes, and make the old names wrappers.

**Comment by kgaillot** [comment]
> > That's exactly what PCMK__NO_COMPAT is for. This particular PR is unusual in that we're keeping the APIs available internally, so it doesn't work here. Probably the cleanest solution is to not deprecate anything until a new private API has replaced them internally, so we can use PCMK__NO_COMPAT. Now that we have coccinelle, that should easy in most cases, so maybe I'll go that route.
> 
> I know but when introducing PCMK__NO_COMPAT in addition wherever the doxygen-comments are that clutters the code and is just useful for such a deprecation test-build. Code would be turned on in the pacemaker-build and off for e.g. test-building sbd to see if deprecated API is still used.

Actually no -- crm_internal.h defines PCMK__NO_COMPAT to ensure that no internal code uses the deprecated stuff. That was its original purpose. The compatibility code is enabled only for external code. That does highlight that we haven't been testing that the compatibility code compiles -- obviously dlm and sbd CI will do it but hopefully we'd catch it earlier than that. :)

**Comment by kgaillot** [comment]
Rebased on current master.

To continue the deprecation discussion, here is how glib handles it:

https://developer.gnome.org/Deprecations/

They define per-release function decorators for deprecations (e.g. GLIB_DEPRECATED_IN_2_26) and introductions (e.g. GLIB_AVAILABLE_IN_2_42), and allow callers to define minimum and maximum releases that should be supported. APIs deprecated before the minimum will cause a compiler warning if used, and APIs introduced after the maximum will not be usable.

That's an interesting approach but I think it's overkill for pacemaker. I think PCMK__NO_COMPAT is good for our purposes, but it's an internal symbol, so if we want to provide external library users similar functionality, I'm thinking of replacing it with this in crm/crm.h:

    #ifndef PCMK_ALLOW_DEPRECATED
    #define PCMK_ALLOW_DEPRECATED 1
    #endif

then crm_internal.h and external callers can define it to 0 before including crm.h to make deprecated code unavailable.

**Comment by clumens** [comment]
I like this approach, and I especially like changing PCMK__NO_COMPAT to PCMK_ALLOW_DEPRECATED.  It gets rid of the double negative when using ifndef which makes things easier to follow.

**Comment by kgaillot** [comment]
Rebased on current master again, and added commit for PCMK_ALLOW_DEPRECATED.

@clumens @wenningerk Is this good to merge?

File: include/crm/crm.h [hunk 2]
```diff
@@ -29,6 +29,21 @@ extern "C" {
...
 
+#ifndef PCMK_ALLOW_DEPRECATED
+/*!
+ * \brief Allow use of deprecated Pacemaker APIs
+ *
+ * By default, external code using Pacemaker headers is allowed to use
+ * deprecated Pacemaker APIs. If PCMK_ALLOW_DEPRECATED is defined to 0 before
+ * including any Pacemaker headers, deprecated APIs will be unusable. It is
+ * strongly recommended to leave this unset for production and release builds,
+ * to avoid breakage when Pacemaker upgrades deprecate more APIs. This should be
```
Comment (clumens) [inline include/crm/crm.h:40]: I think you're missing a word between "upgrades" and "deprecate".  Other than that, fine to merge.

Comment (kgaillot) [reply include/crm/crm.h]: reworded, thanks

**Comment by clumens** [comment]
Ship it!

**Comment by kgaillot** [comment]
retest this please

**Pull request merged**